pkg/util/deeptest/statictesthelper.go

```diff
@@ -0,0 +1,262 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deeptest
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/gobwas/glob"
+	"github.com/stretchr/testify/require"
+)
+
+type staticTestHelper struct {
+	ignorePath         []glob.Glob
+	pointerComparePath []glob.Glob
+}
+
+func (h *staticTestHelper) applyOptions(opts ...Option) {
+	for _, opt := range opts {
+		opt.apply(h)
+	}
+}
+
+func (h *staticTestHelper) shouldIgnore(path string) bool {
+	for _, g := range h.ignorePath {
+		if g.Match(path) {
+			return true
+		}
+	}
+	return false
+}
+
+func (h *staticTestHelper) shouldComparePointer(path string) bool {
+	for _, g := range h.pointerComparePath {
+		if g.Match(path) {
+			return true
+		}
+	}
+	return false
+}
+
+func (h *staticTestHelper) assertRecursivelyNotEqual(t require.TestingT, valA, valB reflect.Value, path string) {
+	if h.shouldIgnore(path) {
+		return
+	}
+
+	if !valA.IsValid() || !valB.IsValid() {
+		require.False(t, !valA.IsValid() && !valB.IsValid(), path+" should not be zero value at the same time")
+		return
+	}
+
+	if valA.Type() != valB.Type() {
```

Comment:
Why not use 	`require.Equal(t, valA.Type(), valB.Type(), path+" should have the same type")`

Reply:
If there are two different implementations for an `interface`, it'd be expected to have different types here. They are also considered not equal.